src/service/resources-impl.js

```diff
@@ -407,6 +410,12 @@ export class Resources {
    * @private
    */
   buildReadyResources_() {
+    // Avoid cases where elements add more elements inside of them
+    // and cause an infinite loop of building - see #3354 for details.
+    if (this.isCurrentlyBuildingPendingResources_) {
+      return;
+    }
+    this.isCurrentlyBuildingPendingResources_ = true;
```

Comment:
Done. Had to break the method into multiple methods to make it a bit more readable.

Let me know what do you think and if I should add some more test cases.